### PR TITLE
Make character limit configurable

### PR DESCRIPTION
**Pull request by lambadalambda**
This adds a character limit configurable by setting and environment variable. It also adds the character limit to the initial state and instance endpoint, so clients can use it to set the correct limit.

Default is still 500 characters if no other value is set.

This relates to #5693 and #4915.

### CODE REVIEW

File: .env.production.sample [hunk 1]
```diff
@@ -134,3 +134,6 @@ STREAMING_CLUSTER_NUM=1
 # If you use Docker, you may want to assign UID/GID manually.
 # UID=1000
 # GID=1000
+
+# Maximum allowed character count
+# MAX_CHARS=500
```
Comment (ThibG) [inline .env.production.sample]: I'd suggest MAX_TOOT_CHARS or something similar, as one may want custom maximum length for other things (such as bios).

**Review by ClearlyClaire — COMMENTED** [review]
Looks good to me, except for the minor config naming issue.

File: .env.production.sample (hunk 1 above)
Comment (lambadalambda) [reply .env.production.sample]: Makes sense, i made that change.

File: .env.production.sample [hunk 2]
```diff
@@ -134,3 +134,6 @@ STREAMING_CLUSTER_NUM=1
 # If you use Docker, you may want to assign UID/GID manually.
 # UID=1000
 # GID=1000
+
+# Maximum allowed character count
+# MAX_TOOT_CHARS=500
```
Comment (m4sk1n) [inline .env.production.sample]: afaik, toots are called statuses in code…

Comment (lambadalambda) [reply .env.production.sample]: You think it's worth to rename it again?

Comment (ThibG) [reply .env.production.sample]: I'm personally fine either ways. `MAX_STATUS_CHARS` would be more consistent with the source code, but the configuration option is intended for admins, not for developers.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by Gargron** [comment]
I have denied such patches in the past. If somebody wants to hack it in, they may do so, but I don't want to encourage it. I feel like the UX is designed, visually and behaviourally, around a certain number of characters, and deviations have a negative impact on it. For example, I want people to pick their Mastodon based on, for example, the policies, or the theme, but I want the base level of functionality to be the same so it doesn't go into the choice.

The same thing as here is with the file size limits. Some people want to make them bigger. Well, great, but if your users make use of those higher limits their attachments/avatars will not show up on other instances which have to cache that stuff. You can do that, but I'm not going to make that an Official Thing.

**Comment by lambadalambda** [comment]
I can understand that, but many instances already change their character counts. Some do it for jokes (e.g. the 512 characters for a tech-oriented instance), some do it for real reasons (the 10.000 characters on an instance for writers).

If it's easily changeable in the main codebase or not isn't really important to me, but I'd like to have some API to get the character count of the instance, so apps can work correctly with modified servers.

**Comment by SirCmpwn** [comment]
>I have denied such patches in the past. If somebody wants to hack it in, they may do so, but I don't want to encourage it. I feel like the UX is designed, visually and behaviourally, around a certain number of characters, and deviations have a negative impact on it. For example, I want people to pick their Mastodon based on, for example, the policies, or the theme, but I want the base level of functionality to be the same so it doesn't go into the choice.

:-1:

In my opinion, this isn't your place to decide - it's for the instances to decide. Don't make them hack it. And they *are* hacking it - I've seen a good handful of toots in my timeline that are rather long, and because Mastodon isn't intelligent about alternative character limits it can make for a very annoying UX to have to scroll past ridiculously long toots.

Mastodon should support arbitrary lengths (but default to 500 is totally fine, opinionated default settings are better than opinionated hardcoded values), and also have a configurable maximum length at which the rest of the toot will be hidden behind a "read more" button. We already have such a UX for content warnings, this isn't breaking new ground design-wise.

Mastodon is good because it gives users that freedom they don't have on Twitter. You should embrace that.

**Comment by connyduck** [comment]
For App-Developers like me it is really important to have an API to get the character limit of the currently connected instance. Please give us an API so we can adapt our clients to the already modified mastodon-servers and provide our users the best experience.
Maybe the limit should not be configurable arbitrarly but only within a sensible limit like 140 -10 000.

**Comment by Lertsenem** [comment]
As a compromise, how about hard coding that toot with more than 500 chars get automatically wrapped in a CW with an ellipsis?

That's already kind of an informal etiquette anyway, and leaves the web UI intact for the most part.

**Comment by akihikodaki** [comment]
I wonder if there is any chance for this pull request to be merged. If so, we can have more discussions and improvements. If not, and this issue gets closed, others who like this change can decide to maintain it by themselves or abandon.

**Comment by lambadalambda** [comment]
@Gargron Can you please either merge or close this?

**Comment by wxcafe** [comment]
Up @Gargron

**Comment by jennifermf** [comment]
can we revisit this? if we can have a billion custom emojo, can we please please please get a 500 character limit bio as the default?

the average mastodon user doesn't fit neatly into twitter-sized boxes. it's time to give people more space to describe who they are.

**Comment by Shreewebs** [comment]
I am looking to increase TOOT character count to 600. It seems file .env.production.sample no longer has such options open.
Looking for a tweak!
Thank you

**Comment by Grawl** [comment]
Just did this at my fork https://github.com/Grawl/mastodon/commit/aef40d7cfa2b27219bf6ef5b08bc804f6fdf7ee2

**Comment by Grawl** [comment]
just did it in my fork https://github.com/Grawl/mastodon/commit/3fd85bcf412235e546fe4ac4dd418f2dc4ddd033

**Comment by tyteen4a03** [comment]
Are there any scripts to work around this?

**Comment by r3d-f0x** [comment]
> As a compromise, how about hard coding that toot with more than 500 chars get automatically wrapped in a CW with an ellipsis?

I would **not** want this done.  Would love to see long toots past the default 500 characters be limited to the first 500 up to the complete word and then expandable by the user, but using "CW" will devalue the use of CW.  I run a right leaning site an intend to use CW as a way of moderating toots that could be problematic without outright deleting them.  Doing anything that would get people into the habit of clicking to see CW without thinking about it would make it a lot more challenging to deal with problematic posts.

**Comment by itbeard** [comment]
Hello! @Gargron maybe in 2022 it makes sense to reconsider your decisions about this feature?

**Comment by rkingett** [comment]
I said this in another issue but I have no problem saying it here. Many Blind instance administrators want this, especially because it is easier to read one lengthy post than 60 posts broken up into a thread and Blind instance admins want to cultivate a space dedicated to their users needs.

**Comment by JiFish** [comment]
I wanted to add another reason for this that I don't think has been mentioned yet: I'm using managed Mastodon hosting for my instance. In this situation, I can't reasonably expect my host to apply a hack for just me.

**Comment by expertmanofficial** [comment]
I would like to wonder how do people comment here even through the issue is already closed?

**Comment by tyteen4a03** [comment]
> I would like to wonder how do people comment here even through the issue is already closed?

Closed issues only mean the issue is considered closed. Maintainers can choose to lock an issue which is another thing.

**Comment by yawnbox** [comment]
1. Please give communities freedom to choose how much their profiles can say, without hacks.
2. Please lower the bar for admins so the diversity of admins increases.

**Comment by rkingett** [comment]
I can't tell if @Gargron  knows why people pick servers or not. People already pick servers based on a higher character limit and on the public timelines I see people asking for two things.

1. A high character limit. 2,000 and up.
2. An instance theme to go along with the high character limit.

As I said previously, the flagship instances can keep this at 500 if they want. Just make it easier for admins to cadre to their communities.

**Comment by Simx72** [comment]
> People already pick servers based on a higher character limit and on the public timelines I see people asking for two things.

I do. Seeing if in 2026, this has changed.

**Pull request closed without merging**